### ISSUE TITLE
huProvBuild

### ISSUE THREAD

**Issue by timm** [researchart/rose6icse#16]
Improving Data Scientist Efficiency with Provenance

https://github.com/researchart/rose6icse/tree/master/submissions/available/huProvBuild
https://github.com/researchart/rose6icse/tree/master/submissions/reusable/huProvBuild

Contact:

- Jingmei Hu
email: jingmei_hu@g.harvard.edu
github ids: CrystalMei

**Comment by minkull** [comment]
## Author
```tex
\author{Jingmei Hu}
\affiliation{
	\institution{Harvard University}
	\streetaddress{33 Oxford St.}
	\city{Cambridge}
	\state{MA}
	\postcode{02138}
	\country{USA}
}
  	
\author{Jiwon Joung}
\affiliation{
  \institution{Harvard University}
  \streetaddress{33 Oxford St.}
  \city{Cambridge}
  \state{MA}
  \postcode{02138}
  \country{USA}
}
	
\author{Maia Jacobs}
\affiliation{
	\institution{Harvard University}
	\streetaddress{33 Oxford St.}
	\city{Cambridge}
	\state{MA}
	\postcode{02138}
	\country{USA}
}
	
\author{Margo I. Seltzer}
\affiliation{
  \institution{University of British Columbia}
  \streetaddress{201-2366 Main Mall}
  \city{Vancouver}
  \state{BC}
  \postcode{V6T 1Z4}
  \country{Canada}
}
	
\author{Krzysztof Gajos}
\affiliation{
	\institution{Harvard University}
	\streetaddress{33 Oxford St.}
	\city{Cambridge}
	\state{MA}
	\postcode{02138}
	\country{USA}
}
```

**Comment by artifacts-reviewer** [comment]
I was able to use the web application as described in the README file. It would be useful to explain in a bit more detail in the README what output is expected where searching for variables (for example, variable e). I found the terminal output of "The variable name is e ..." a bit confusing. Some text that explains this and what is expected in the ProvScript.py box would be useful.

In addition, the README only explains how to use ProvBuild with the test file. It would be good to make the README more general and use the test file to show a specific example.

I also suggest adding a brief description of ProvBuild in the README (even if it is just the abstract of the ICSE paper) rather than just refer to the ICSE paper title.

**Comment by anonreporev** [comment]
Hello,

I was able to see the web interface after running python app.py.

When entering Test.py into either file box, then clicking submit, I get:
Internal Server Error
The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application.

The python prompt shows
 "File "app.py", line 57, in normal
    file = open("record-time/session.txt", "w")
IOError: [Errno 2] No such file or directory: 'record-time/session.txt'
127.0.0.1 - - [10/Feb/2020 16:20:45] "[35m[1mPOST /normal HTTP/1.1[0m" 500 -"

I tried also Demo.py with the same error.

It's also not clear to me if I should be using ProvBuild Mode or Normal Mode, and what is the purpose of user name?

Thanks.

**Comment by jingmeihu** [comment]
> Hello,
> 
> I was able to see the web interface after running python app.py.
> 
> When entering Test.py into either file box, then clicking submit, I get:
> Internal Server Error
> The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application.
> 
> The python prompt shows
> "File "app.py", line 57, in normal
> file = open("record-time/session.txt", "w")
> IOError: [Errno 2] No such file or directory: 'record-time/session.txt'
> 127.0.0.1 - - [10/Feb/2020 16:20:45] "�[35m�[1mPOST /normal HTTP/1.1�[0m" 500 -"
> 
> I tried also Demo.py with the same error.
> 
> It's also not clear to me if I should be using ProvBuild Mode or Normal Mode, and what is the purpose of user name?
> 
> Thanks.

ProvBuild and Normal mode should both work without logging times.
The records under record-time and user name settings are designed for user study 1 in the paper. It needs to collect completion time and correctness of the tasks.

**Comment by anonreporev** [comment]
@CrystalMei 
OK, but if I get this error, it means I can't run and test your application.

Should the file be changed so it can be run?

**Comment by jingmeihu** [comment]
> I was able to use the web application as described in the README file. It would be useful to explain in a bit more detail in the README what output is expected where searching for variables (for example, variable e). I found the terminal output of "The variable name is e ..." a bit confusing. Some text that explains this and what is expected in the ProvScript.py box would be useful.
> 
> In addition, the README only explains how to use ProvBuild with the test file. It would be good to make the README more general and use the test file to show a specific example.
> 
> I also suggest adding a brief description of ProvBuild in the README (even if it is just the abstract of the ICSE paper) rather than just refer to the ICSE paper title.

I updated the Readme files in those:
Artifact Evaluation repo: [https://github.com/CrystalMei/rose6icse](https://github.com/CrystalMei/rose6icse)
ProvBuild repo: [https://github.com/CrystalMei/ProvBuild](https://github.com/CrystalMei/ProvBuild)

**Comment by jingmeihu** [comment]
> @CrystalMei
> OK, but if I get this error, it means I can't run and test your application.
> 
> Should the file be changed so it can be run?

I cannot reproduce your error on my side, so I would suggest that you could try to create the following three folders under the main `ProvBuild` folder:
- record-time
- Task-results
- Task-time

Hope it will work!
I also updated the Readme file with a Demo: [https://github.com/CrystalMei/ProvBuild/blob/master/README.md](https://github.com/CrystalMei/ProvBuild/blob/master/README.md)

**Comment by anonreporev** [comment]
@CrystalMei    I also don't know why I'm getting this error and and others are not.

I have Python 2.7.17 on Windows 10.  I've downloaded all the listed packages with pip.

It's not my role to debug your code, but there is obviously a scenario where the record-time directory is opened but not created.  

The problem is that this error may occur for others that use your app, so if it's not understood and fixed, the artefact may not be usable for some users.

Your fix may work for me, but what about any further users with the same problem?

**Comment by jingmeihu** [comment]
> @CrystalMei I also don't know why I'm getting this error and and others are not.
> 
> I have Python 2.7.17 on Windows 10. I've downloaded all the listed packages with pip.
> 
> It's not my role to debug your code, but there is obviously a scenario where the record-time directory is opened but not created.
> 
> The problem is that this error may occur for others that use your app, so if it's not understood and fixed, the artefact may not be usable for some users.
> 
> Your fix may work for me, but what about any further users with the same problem?

The reason for your error is because Windows and Linux/macOS have different command support (e.g. `rm` and fold slash and backslash). 
I removed related code in `app.py` for generality. ProvBuild and all tests are developed under Linux and macOS, and if you want to run ProvBuild on Windows 10, please use [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/install-win10).

I also update [README](https://github.com/CrystalMei/ProvBuild/blob/master/README.md) about "Windows may not be supported by ProvBuild"

**Comment by artifacts-reviewer** [comment]
Everything works fine for me on a mac and I am happy with the updates in the README.

I suggest one additional update to clarify that a username is not required and can be left blank (also note why that text box is there otherwise it will be a bit confusing).

**Comment by jingmeihu** [comment]
> Everything works fine for me on a mac and I am happy with the updates in the README.
> 
> I suggest one additional update to clarify that a username is not required and can be left blank (also note why that text box is there otherwise it will be a bit confusing).

Thanks for your suggestions!
I updated the Readme files in those:
Artifact Evaluation repo: [https://github.com/CrystalMei/rose6icse/tree/master/submissions/available/huProvBuild](https://github.com/CrystalMei/rose6icse/tree/master/submissions/available/huProvBuild)
[https://github.com/CrystalMei/rose6icse/tree/master/submissions/reusable/huProvBuild](https://github.com/CrystalMei/rose6icse/tree/master/submissions/reusable/huProvBuild)
ProvBuild repo: https://github.com/CrystalMei/ProvBuild

**Comment by artifacts-reviewer** [comment]
Thanks. I think the files now well describe the artifacts. To obtain the available and reusable badges, I suggest the authors to upload all files into an online archive such as Zenodo or Figshare so a DOI can be obtained.

**Comment by jingmeihu** [comment]
> Thanks. I think the files now well describe the artifacts. To obtain the available and reusable badges, I suggest the authors to upload all files into an online archive such as Zenodo or Figshare so a DOI can be obtained.

The submitted paper contained an anonymous Zenodo link and the camera-ready paper contained the final Zenodo link as a reference.

**Comment by artifacts-reviewer** [comment]
Please ensure the zenodo data is updated with the modified files and provide a link here.

**Comment by jingmeihu** [comment]
> Please ensure the zenodo data is updated with the modified files and provide a link here.

Here is the link:
[https://zenodo.org/record/3668377](https://zenodo.org/record/3668377)

**Comment by anonreporev** [comment]
I'm satisfied with the README being updated to say that Windows is not supported.

However, the readme in the main repo was updated: https://github.com/CrystalMei/ProvBuild/blob/master/README.md
I'm not sure if the available and reusable submission's readme also has to be updated or if this is just for us to review.
https://github.com/researchart/rose6icse/blob/master/submissions/reusable/huProvBuild/README.md
https://github.com/researchart/rose6icse/blob/master/submissions/available/huProvBuild/README.md
I'm not sure finally what is linked, your repo, the submitted repos, or the zenodo link?
Otherwise it's fine.

**Comment by jingmeihu** [comment]
I kept my repo and zenodo link up to date. I'm not sure whether I can still update the submitted repos.
Already submitted another pull request #171 to this repo.

**Comment by timm** [comment]
dear @CrystalMei : thank you so much for your responsiveness to queries.

dear @anonreporev @artifacts-reviewer : thanks for your hard work on this one. please confirm your current beliefs regarding what else needs to be done before (or of) we can say "available" and/or "reusable"

**Comment by artifacts-reviewer** [comment]
I believe the badges should be awarded.

**Comment by anonreporev** [comment]
I agree, please award the badges.
